Parallel I/O output module enhancements

SeanBryan51 opened the pull request:
<!-- readthedocs-preview cable start -->
----
📚 Documentation preview 📚: https://cable--655.org.readthedocs.build/en/655/

<!-- readthedocs-preview cable end -->

SeanBryan51 committed:
Add aggregator implementation

Co-authored-by: Lachlan Whyborn <lachlan.s.whyborn@gmail.com>

SeanBryan51 committed:
Add parallel I/O output module implementation

SeanBryan51 committed:
src/offline/cable_define_types.F90: add tscrn_max_daily and tscrn_min_daily aggregators to canopy_type

SeanBryan51 committed:
src/offline/cable_serial.F90: update tscrn_max_daily and tscrn_min_daily aggregators

ccarouge commented on src/util/aggregator_types.F90:
```diff
@@ -0,0 +1,418 @@
+
+    if (method == "mean") then
+      this%accumulate => sum_accumulate
+      this%normalise => mean_normalise
```
I wonder if it would be best to use an incremental average algorithm: https://math.stackexchange.com/questions/106700/incremental-averaging

This way, the accumulate method calculates the mean at each pass and we don't need the special mean_normalise.

SeanBryan51 replied:
Happy to add this in

SeanBryan51 replied:
Incremental averaging added: [fb4febe](https://github.com/CABLE-LSM/CABLE/pull/655/commits/fb4febe5e0fd41793dc5881cca0e5d078140325b)

SeanBryan51 replied:
`normalise` procedure removed: [5708824](https://github.com/CABLE-LSM/CABLE/pull/655/commits/570882442a3d0219005f075c2c98c7d3df158bbb)

ccarouge commented on src/util/aggregator_types.F90:
```diff
@@ -0,0 +1,418 @@
+      this%reset => other_reset
+    elseif (method == "point") then
+      this%accumulate => point_accumulate
+      this%normalise => other_normalise
```
This does not correspond to the subroutines it seems. In the subroutines, you have `point_accumulate` that doesn't do anything and `point_normalise` that attributes the value at the timestep.

I would rather do it as you have it here: `point_accumulate` attributes the current values, then`other_normalise` does nothing.

If we go with an incremental average, then we don't need any normalise step (all the aggregations use the `other_normalise` subroutine that doesn't do anything.

ccarouge commented on src/offline/cable_output_prototype_v2.F90:
```diff
@@ -0,0 +1,713 @@
+          call aggregator_handle%accumulate()
+        end associate
+      end do
+    end if
```
I understand I think. Is it for CASA variables that are only updated daily?
So we would have an equivalent one for yearly accumulation for BLAZE for example?

Whyborn commented on src/offline/cable_output_definitions.F90:
```diff
@@ -0,0 +1,228 @@
+      active=output%Qh .and. (output%patch .OR. patchout%Qh), &
+      decomp=output_decomp_base_patch_real32, &
+      aggregator=new_aggregator( &
+        source_data=canopy%fh, &
```
I'm actually not 100% sure what happens if you specify a NetCDF variable to be 64 bit float, but provide a 32 bit array- I assume NetCDF (PIO) does the conversion?

SeanBryan51 replied:
The NUG describes the [type conversion behaviour](https://docs.unidata.ucar.edu/nug/current/md_types.html#:~:text=If%20the%20type,the%201024th%20power.) of netCDF.

SeanBryan51 commented on src/offline/cable_serial.F90:
```diff
@@ -572,6 +588,17 @@ SUBROUTINE serialdrv(NRRRR, dels, koffset, kend, GSWP_MID, PLUME, CRU, site, mpi
+                start_year = CABLE_USER%YearStart
+            ENDIF
+          END IF
+
```
> Not sure that's what you mean in your TODO. It seems to apply more to the code in cable_timing_utils.F90 than here.

You've interpreted that correctly, it was a reminder to myself to look into an alternative algorithm for monthly timings that doesn't require the start year of the simulation.

SeanBryan51 replied:
> What's the reason not to have `cable_user%yearstart` be something that is unconditionally required in the namelist, and have that be the final truth about the start point of the simulation? Having that able to be overwritten is bound to lead to confusion in the future.

That's a great point, it might make sense to instead error if `met%year(1)` does not match `cable_user%yearstart`, and rely on `cable_user%yearstart` for timing logic.

I don't really understand why `met%year(1)` is required for this case, I'll see what happens when we rely on `cable_user%yearstart` alone

Whyborn replied:
Isn't `ktau` already that? The timestep within the current year, with `ktau_tot` being the total timestep in the simulation?

SeanBryan51 commented:
Thanks for the comments @ccarouge, I'll add in those suggestions!

ccarouge commented on src/offline/cable_output_prototype_v2.F90:
```diff
@@ -0,0 +1,713 @@
+      end do
+    end if
+
+    if (time_step_matches(dels, time_index, output%averaging, leaps, start_year)) then
```
I would split this subroutine in two (or more) to split out where we accumulate the aggregators and where we write out the data. I think we should have an update and a write procedure. I've seen you discussed this with Lachlan as well. I understand the idea of replicating the current behaviour. I'm just worried this is going to push us towards a design we would have to rethink to split to get the accumulation step handled by the science modules.

SeanBryan51 replied:
Lachlan and I sort of came to the agreement that any aggregators used as part of non-output related computations would be driven outside the output module (similar to the `canopy%tscrn_max_daily` aggregator), and that aggregators for writing output would be driven by the output module. Is there a specific use case you have in mind where it makes sense to have the accumulation step for the aggregators for writing output handled by the science modules?

Whyborn replied:
It doesn't have to go to one of the science modules. I think it still makes sense to accumulate and write in different routines.

SeanBryan51 commented on src/util/aggregator_types.F90:
```diff
@@ -0,0 +1,418 @@
+  end type aggregator_handle_t
+
+  type, extends(aggregator_t) :: aggregator_int32_1d_t
+    integer(kind=int32), dimension(:), allocatable :: storage
```
Added in [676e718](https://github.com/CABLE-LSM/CABLE/pull/655/commits/676e718d599af187a6a84bfc425329e684c7288e)

ccarouge commented on src/offline/cable_output_prototype_v2.F90:
```diff
@@ -0,0 +1,713 @@
+
+  end subroutine write_variable
+
+  subroutine write_variable_grid_cell_average(output_variable, output_file, time_index, patch, landpt)
```
@Whyborn , my first comment was to say I'd prefer the grid averaging to be done separately from the write. But then, I thought Sean must have had a reason to do it this way. I thought hard (and not that long) and thought the memory thing could be it.

SeanBryan51 replied:
> I don't understand why we need to save the averaging within output_variable%, why not use local variables to store the data for the write?

Just to confirm, are you asking whether the time aggregated quantities should be the grid cell averaged values rather than the values per patch?

Whyborn replied:
@ccarouge I think it's actually quite difficult to decouple them completely, because the grid cell averages don't have working variables to be associated with.

SeanBryan51 commented on src/offline/cable_output_prototype_v2.F90:
```diff
@@ -0,0 +1,713 @@
+      end do
+    end if
+
+    if (time_step_matches(dels, time_index, output%averaging, leaps, start_year)) then
```
> Note, we need to change "output%averaging" to "output%aggregating" since it's not always an average on that time window.

Just thought of this now, a way to make this a bit nicer is to have an `aggregation_frequency` component in the `cable_output_profile_t` type which is assigned to `output%averaging` on initialisation. I think this makes sense as the sampling frequency is intended to be set at the profile level.

SeanBryan51 committed:
Introduce cable_enum_t for improved type safety for enum data structures

SeanBryan51 committed:
src/util/cable_timing_utils.F90: Implement case for custom hourly frequencies

SeanBryan51 committed:
src/offline/cable_mpimaster.F90: Accumulate tscrn aggregators in master driver

SeanBryan51 committed:
src/util/io/output/cable_output_core.F90: fix time axis values for 'all' case

Whyborn commented:
@SeanBryan51 I'm going to take a look at this tomorrow- where would you recommend I start? Is there a logical place to start?

SeanBryan51 commented:
I would say the commits from and including [Add aggregator implementation](https://github.com/CABLE-LSM/CABLE/pull/655/commits/3726642bdb9a96e7725e97806c90a7f615619888) would be relevant for review - commits prior are less relevant IMO.

I.e: https://github.com/CABLE-LSM/CABLE/compare/129d746...parallelio-playground-output-module-enhancements

Whyborn commented on src/util/io/output/cable_output_utils.F90:
```diff
@@ -0,0 +1,960 @@
+  public :: associate_temp_buffer_real32
+  public :: associate_temp_buffer_real64
+
+  ! Decomposition pointers for each variable class and data type
```
Why are these defined here? Why don't we just define them when we use them, when we actually define the output variables? This hides information that will be necessary for future developers. If the decompositions are defined in the same location as the output variables, then all the information new developers need is in one place

SeanBryan51 replied:
I've made a pretty significant change in [0e594f5](https://github.com/CABLE-LSM/CABLE/pull/655/commits/0e594f51aa7779fc287a88fe56d78c31590dd792), I split up `cable_output_utils_mod` into three modules, namely:
1. `cable_output_utils_mod`
2. `cable_output_decomp_mod`
3. `cable_output_reduction_buffers_mod`

I also followed your advice on defining the decompositions with the output module code, rather associating pointers to a shared `io_decomp` derived type - I think this makes it easier to understand where the decompositions are being defined.

Whyborn commented on src/util/cable_enum.F90:
Seems to be missing the core functionality of an enum, which is that it only has specified allow values

Whyborn replied:
Could we have a vc sometime next week so you can explain what the benefit of this is and how it works? Cause I'm still not getting why it's important to do this way.

Whyborn commented on src/util/io/output/cable_output.F90:
```diff
@@ -0,0 +1,26 @@
+  use cable_output_types_mod, only: CABLE_OUTPUT_DIM_SNOW
+  use cable_output_types_mod, only: CABLE_OUTPUT_DIM_RAD
+  use cable_output_types_mod, only: CABLE_OUTPUT_DIM_PLANTCARBON
+  use cable_output_types_mod, only: CABLE_OUTPUT_DIM_SOILCARBON
```
We already have these values defined elsewhere in `science/casa-cnp/casa_dimension.F90` for the CASA stuff, and `params/grid_constants_cbl.F90` for the others.

Whyborn reviewed (commented):
First pass at the review, I think I'll have to go through everything again to get it right in my head I think

SeanBryan51 commented on src/util/io/output/cable_output_types.F90:
```diff
@@ -0,0 +1,59 @@
+    real, dimension(2) :: range
+    type(cable_output_dim_t), allocatable :: data_shape(:)
+    class(aggregator_t), allocatable :: aggregator
+  end type
```
> Did you have a plan for how to turn entries read from the output configuration file into `cable_output_variable`s?

It should be pretty easy I think. All diagnostics/fields are "registered" in memory, so all we have to do is discard the fields that weren't requested for a stream/profile and initialise the ones that were. We can also instantiate multiple copies of each `cable_output_variable_t` when we have variants of the same field (e.g. different reduction_method or aggregation_method).

SeanBryan51 replied:
I think the intention is to unify the restart and output formats in the future, which would actually remove a lot of the `if (restart)` complexity in the code. I'll mark the rest of the discussion here as "resolved" for now

SeanBryan51 committed:
src/util/io/output/cable_output_utils.F90: remove unused variable

SeanBryan51 committed:
src/offline/cable_mpimaster.F90: Add output module to MPI master driver

SeanBryan51 committed:
src/offline/cable_serial.F90: fix guarding if condition for output module finalisation

SeanBryan51 commented on src/offline/cable_serial.F90:
```diff
@@ -1004,16 +1044,20 @@ SUBROUTINE serialdrv(NRRRR, dels, koffset, kend, GSWP_MID, PLUME, CRU, site, mpi
     IF (cable_user%CALL_climate) &
       CALL WRITE_CLIMATE_RESTART_NC ( climate, ktauday )
 
+    call cable_output_mod_end()
```
Addressed in [64b3cf4](https://github.com/CABLE-LSM/CABLE/pull/655/commits/64b3cf496be4d0e70417437dc0ac34c62420aa10)

SeanBryan51 commented on src/util/io/output/cable_output_core.F90:
```diff
@@ -0,0 +1,789 @@
+          accumulation_frequency=output_var%accumulation_frequency, &
+          var_name=output_var%name, &
+          file_name=global_profile%file_name &
+        )
```
> Is that the definition of all possible output variables? I don't see how that could possibly be the case, due to the possible permutations of frequency, aggregation and reduction.

Yes you're right that it wouldn't be all possible output variables - just to be clear, I don't want to confuse this work with output redesign proposal, which would introduce the ability to specify aggregation and reduction methods on a per variable basis. In this work, variables with different aggregation methods are registered independently, as this is not currently configurable.

With the new output redesign proposal, the intention for the `registered_output_variables` list is to have a 1-to-1 relationship with all "working variables" which are used by the output module. This has a couple implications, namely: 1. each registered output variable entry would not specify an aggregation and reduction method (left undefined as this comes from the output configuration file), and 2. each registered output variable in `registered_output_variables` is unique in it's `name` (i.e. no duplicates). When an output variable is requested from the output configuration file, we loop over the `registered_output_variables` list and find the entry matching against the requested `name` component - once the entry is found, the entry is copied into the list of output variables for the current stream using Fortran's intrinsic assignment, here we also define the requested reduction and aggregation method. How does that sound?

Whyborn replied:
So the current design is strictly to allow it to work with the current way of managing CABLE output, through the namelist options? And then integrating the new output configuration file would come later?

SeanBryan51 replied:
Yes

SeanBryan51 committed:
src/util/io/output/cable_output_core.F90: fix non-standard array constructor syntax for strings

SeanBryan51 committed:
src/util/io/output/cable_output_utils.F90: fix non-standard array constructor syntax for strings

SeanBryan51 committed:
src/util/io/output/cable_output_utils.F90: bug fix in infer_dim_names procedure

SeanBryan51 committed:
src/util/io/output/cable_output_utils.F90: bug fix for comparing data_shape arrays

SeanBryan51 committed:
src/util/io/output/cable_output_utils.F90: fix cell_methods attribute

SeanBryan51 committed:
src/util/io/output/cable_output_core.F90: Remove redundant temp_buffer_* pointers

SeanBryan51 committed:
src/util/io/output/cable_output_core.F90: Remove potential for implicit save

SeanBryan51 committed:
Add variable and global attributes list

SeanBryan51 committed:
Split out decomposition and reduction buffer functionality from cable_output_utils_mod

Introduce `grid_type` component and replace grid based `if (restart)`
with `select case(grid_type)`

SeanBryan51 committed:
Remove cable_io_decomp_mod

SeanBryan51 committed:
Clean up sampling frequency check subroutine

SeanBryan51 committed:
src/util/io/output/cable_output_definitions.F90: remove restart argument

SeanBryan51 committed:
src/util/aggregator.F90: Add scale and offset procedures

SeanBryan51 committed:
src/util/aggregator.F90: Add aggregator_type procedure

SeanBryan51 committed:
Introduce diagnostic variables for CABLE outputs

Note that some biogeophysics diagnostic variables overlap with CASA-CNP
diagnostic variables, e.g. `canopy%fgpp` and `canopy%fnpp`. When
CASA-CNP is enabled, these diagnostic variables are re-calculated using
CASA-CNP variables before writing the diagnostic variables to the output
file. The overriding of diagnostic variables is done to match the
current behaviour of the output module - there may be scope to enable
both CASA and non-CASA variants of these diagnostics in the future.

SeanBryan51 committed:
Remove duplicate namelist variables

SeanBryan51 committed:
Add all output and restart definitions, bug fixes and other quality of life changes

In src/offline/cable_mpimaster.F90 and src/offline/cable_serial.F90,
daily aggregators are reset immediately after writing instead of in the
next iteration as this breaks for the last ktau iteration.

New components have been added to `cable_output_variable_t`. These
include `scale`, `offset` which are useful for applying unit
conversions, and also `field_name` and `netcdf_name` to facilitate both
restart variable names (`field_name`) and output variable names
(`netcdf_name`).

All components of the cable_output_variable_t except `field_name` and
`aggregator` now have defaults. This cleans up the interface for
defining output variables.

Improved validation checks on registered output variables.

A new `coordinate_variables` component has been added to
`cable_output_profile_t` such that coordinate variables can be treated
differently from the requested outputs due to differing metadata
requirements.

Removed start_year hack from drivers by overriding cable_user%yearstart
to syear for the default MetType case.

SeanBryan51 committed:
Add carbon molar mass physical constant

SeanBryan51 committed:
src/util/cable_timing_utils.F90: Add seconds_per_day constant

SeanBryan51 committed:
Split output variable declarations into multiple modules

SeanBryan51 commented:
Closing as this work has been superseded by #706, #712 and #713
